modules/etcd/etcd.go

```diff
@@ -0,0 +1,279 @@
+// For the cluster, the first node creates the cluster and the other nodes join it as child nodes.
+type EtcdContainer struct {
+	testcontainers.Container
+	// childNodes contains the child nodes of the current node, forming a cluster
+	childNodes []*EtcdContainer
+	opts       options
+}
+
+// Terminate terminates the etcd container, its child nodes, and the network in which the cluster is running
+// to communicate between the nodes.
+func (c *EtcdContainer) Terminate(ctx context.Context) error {
+	var errs []error
+
+	// child nodes has no other children
+	for i, child := range c.childNodes {
+		if err := child.Terminate(ctx); err != nil {
+			errs = append(errs, fmt.Errorf("terminate child node(%d): %w", i, err))
+		}
+	}
+
+	if c.Container != nil {
+		if err := c.Container.Terminate(ctx); err != nil {
+			errs = append(errs, fmt.Errorf("terminate cluster node: %w", err))
+		}
+	}
+
+	// remove the cluster network if it was created, but only for the first node
+	// we could check if the current node is the first one (index 0),
+	// and/or check that there are no child nodes
+	if c.opts.clusterNetwork != nil && c.opts.currentNode == 0 {
+		if err := c.opts.clusterNetwork.Remove(ctx); err != nil {
+			errs = append(errs, fmt.Errorf("remove cluster network: %w", err))
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+// Run creates an instance of the etcd container type
+func Run(ctx context.Context, img string, opts ...testcontainers.ContainerCustomizer) (*EtcdContainer, error) {
+	req := testcontainers.ContainerRequest{
+		Image: img,
+		Cmd:   []string{},
+	}
+
+	genericContainerReq := testcontainers.GenericContainerRequest{
+		ContainerRequest: req,
+		Started:          true,
+	}
+
+	settings := defaultOptions(&req)
+	for _, opt := range opts {
+		if apply, ok := opt.(Option); ok {
+			apply(&settings)
+		}
+		if err := opt.Customize(&genericContainerReq); err != nil {
+			return nil, err
+		}
+	}
+
+	clusterOpts, err := configureCluster(ctx, &settings, opts)
+	if err != nil {
+		return nil, fmt.Errorf("configure cluster: %w", err)
+	}
+
+	// configure CMD with the nodes
+	genericContainerReq.Cmd = configureCMD(settings)
+
+	// Initialise the etcd container with the current settings.
+	// The cluster network, if needed, is already part of the settings,
+	// so the following error handling returns a partially initialised container,
+	// allowing the caller to clean up the resources with the Terminate method.
+	c := &EtcdContainer{opts: settings}
+
+	if settings.clusterNetwork != nil {
+		// apply the network to the current node
+		err := tcnetwork.WithNetwork([]string{settings.nodeNames[settings.currentNode]}, settings.clusterNetwork)(&genericContainerReq)
+		if err != nil {
+			return c, fmt.Errorf("with network: %w", err)
+		}
+	}
+
+	container, err := testcontainers.GenericContainer(ctx, genericContainerReq)
+	if container != nil {
+		c.Container = container
+	}
```

Comment:
suggestion: we don't need the check any more
```suggestion
	c.Container = container
```

Reply:
It kind of does, but the changes already made mean that it doesn't make sense any more.

The original would have looked like the following, we only needed the `if` because we want to return a `nil` if `container` was `nil` so `CleanupContainer` works. This not the case anymore because we have other clean up we need to handle, so we always return an initialised `EtcContainer` and have the overridden `Terminate` method handle the if `c.Container` check.
```go
	var c *EtcdContainer
	if container != nil {
		c = &EtcdContainer{Container: container}
	}
```

Actually you could just make it:
```go
	c.Container, err = testcontainers.GenericContainer(ctx, genericContainerReq)
```

Reply:
oh I now know what you mean; I think it's more meaningful and readable for eventual readers to have the container-nil-check in the Run function than in the terminate method: we could add the check in the latter too for being defensive, but having in the "constructor" function makes it more understandable.

Wdyt?

Reply:
I think its more confusing having the `if` check given `c.Container` is `nil` by default so has to be handled by `Terminate` anyway, so I would just use:
```go
	if c.Container, err = testcontainers.GenericContainer(ctx, genericContainerReq); err != nil {
		return c, fmt.Errorf("generic container: %w", err)
	}
```

But not a blocker.

Reply:
Well, it actually does not matter, as the c.Container will be nil anyway. Changing it as you suggested, thanks!